impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+    private BaseClient tokenClient;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration apiClientConfiguration) {
+        Assert.notNull(apiClientConfiguration, "apiClientConfiguration must not be null.");
+        ClientConfiguration tokenClientConfig = constructTokenClientConfig(apiClientConfiguration);
+        this.tokenClient = new BaseClient(tokenClientConfig, Caches.newDisabledCacheManager()) {};
+        this.tokenClientConfiguration = tokenClientConfig;
+    }
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration apiClientConfiguration, BaseClient tokenClient) {
+        Assert.notNull(apiClientConfiguration, "apiClientConfiguration must not be null.");
+        Assert.notNull(tokenClient, "tokenClient must not be null.");
+        this.tokenClient = tokenClient;
+        ClientConfiguration tokenClientConfig = constructTokenClientConfig(apiClientConfiguration);
+        this.tokenClientConfiguration = tokenClientConfig;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken() throws IOException, InvalidKeyException, OAuth2TokenRetrieverException {
+        log.info("Attempting to get OAuth2 access token for client id {} from {}",
+            tokenClientConfiguration.getClientId(), tokenClientConfiguration.getBaseUrl() + TOKEN_URI);
+
+        String signedJwt = createSignedJWT();
+        String scope = String.join(" ", tokenClientConfiguration.getScopes());
+
+        try {
+            ExtensibleResource accessTokenResponse = tokenClient.http()
+                .addHeaderParameter("Accept", MediaType.APPLICATION_JSON_VALUE)
+                .addHeaderParameter("Content-Type", MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+                .addQueryParameter("grant_type", "client_credentials")
+                .addQueryParameter("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+                .addQueryParameter("client_assertion", signedJwt)
+                .addQueryParameter("scope", scope)
+                .post(tokenClientConfiguration.getBaseUrl() + TOKEN_URI, ExtensibleResource.class);
+
+            OAuth2AccessToken oAuth2AccessToken = new OAuth2AccessToken();
+            oAuth2AccessToken.setTokenType(accessTokenResponse.getString(TOKEN_TYPE_KEY));
+            oAuth2AccessToken.setExpiresIn(accessTokenResponse.getInteger(EXPIRES_IN_KEY));
+            oAuth2AccessToken.setAccessToken(accessTokenResponse.getString(ACCESS_TOKEN_KEY));
+            oAuth2AccessToken.setScope(accessTokenResponse.getString(SCOPE_KEY));
+
+            log.info("Got OAuth2 access token for client id {} from {}",
+                tokenClientConfiguration.getClientId(), tokenClientConfiguration.getBaseUrl() + TOKEN_URI);
+
+            return oAuth2AccessToken;
+        } catch (Exception e) {
+            log.error("Exception occurred while trying to get OAuth2 access token for client id {}",
```

Comment:
nit: this could be log spam, in the case of a successful retry attempt (as in there was a somewhat expected error, and the application recovered correctly).  This could be moved into the retryable block below as a "debug" message 🤔

Reply:
done.